### ISSUE TITLE
VSCode opens a background Julia process that's constantly using 100% CPU for days

### ISSUE THREAD

**Issue by NHDaly** [julia-vscode/julia-vscode#1093]
Whenever I open VSCode, I constantly have one or sometimes two julia processes that are each using nearly 100% cpu, but i don't know why. :cry:

I realllly want to switch to VSCode because it seems so wonderful! :slightly_smiling_face: So many aspects are better than Atom, including their Live Share, the terminal integration, etc, and the julia support seems really fantastic! I'm liking the linter experience so far. :slightly_smiling_face: So thanks for that!

----------

I'm not sure if it's related to having an internal `.julia` directory that the indexing service is going wild indexing, but we _think_ we addressed that in #1085. 

Notably, i do _not_ see indexing UI in the status bar, which @davidanthoff said in slack is a strong indicator that it's _not_ the indexing service using all that CPU.

I'm trying to debug this now.

OS: MacOS.
Version: Julia Insider v0.15.18
(But i saw this on the standard Julia v0.15.18 as well)

**Comment by NHDaly** [comment]
This all looks like good evidence that it's the LS:


Under "open files and ports" in activity monitor:
```
cwd
/Users/daly/.vscode/extensions/julialang.language-julia-insider-0.15.18/scripts/languageserver
```
Parent process: `Code Helper (Renderer) (38976)`

And here are the command line arguments for the running process:
```julia
ps auxww | grep julia
daly             39439 100.0  9.2  6354144 1543188   ??  R     7:50AM   7:39.98 /Users/daly/bin/julia --startup-file=no --history-file=no --project=/Users/dal
y/.vscode/extensions/julialang.language-julia-insider-0.15.18/scripts/languageserver/packages main.jl /Users/daly/Documents/work/rai/raicode2 --debug=no /var/
folders/l7/4z33k_6j343f9kfpsj4545g00000gn/T/vscode-language-julia-lscrashreports-38976
```

**Comment by NHDaly** [comment]
@davidanthoff thanks for the tips! I set `julia.trace.server` to `verbose`, and indeed i think that this shows the issue:

The resultant `Julia Language Server trace` shows that the LS is spending a bunch of time linting / discovering issues in all the files in the `./.julia` directory, which i had tried to exclude via `"files.exclude"` in https://github.com/julia-vscode/julia-vscode/issues/1085:

```julia
[Trace - 8:01:19 AM] Received notification 'textDocument/publishDiagnostics'.
Params: {
    "uri": "file:///Users/daly/Documents/work/rai/raicode2/.julia/packages/EzXML/QtGgF/docs/setup.jl",
    "version": 0,
    "diagnostics": [
        {
            "range": {
                "start": {
                    "line": 0,
                    "character": 6
                },
                "end": {
                    "line": 0,
                    "character": 9
                }
            },
            "severity": 2,
            "code": "Julia",
            "source": "Julia",
            "message": "Missing reference: Pkg",
            "relatedInformation": null
        },
        {
            "range": {
                "start": {
                    "line": 1,
                    "character": 0
                },
                "end": {
                    "line": 1,
                    "character": 3
                }
            },
            "severity": 2,
            "code": "Julia",
            "source": "Julia",
            "message": "Missing reference: Pkg",
            "relatedInformation": null
        },
        {
            "range": {
                "start": {
                    "line": 2,
                    "character": 0
                },
                "end": {
                    "line": 2,
                    "character": 3
                }
            },
            "severity": 2,
            "code": "Julia",
            "source": "Julia",
            "message": "Missing reference: Pkg",
            "relatedInformation": null
        },
        {
            "range": {
                "start": {
                    "line": 2,
                    "character": 12
                },
                "end": {
                    "line": 2,
                    "character": 23
                }
            },
            "severity": 2,
            "code": "Julia",
            "source": "Julia",
            "message": "Missing reference: PackageSpec",
            "relatedInformation": null
        },
        {
            "range": {
                "start": {
                    "line": 3,
                    "character": 0
                },
                "end": {
                    "line": 3,
                    "character": 3
                }
            },
            "severity": 2,
            "code": "Julia",
            "source": "Julia",
            "message": "Missing reference: Pkg",
            "relatedInformation": null
        }
    ]
}
```

Finally, I moved the `./.julia` directory out (and up one directory), then reloaded the window (and manually killed the orphaned `julia` process it leaves behind), and indeed after waiting less than a minute it finished linting the cwd and the CPU usage disappeared. :)

So i think this is actually just related to #1085: the Language Server doesn't seem to use the `"files.exclude"` flag. :)

**Comment by davidanthoff** [comment]
Ah, yes! So here is some more background info: the symbol server process is in charge of indexing things that your Project.toml points to. The language server process itself, though, will parse all the files inside the folders that are open in VS Code. If that is very large, I guess it can take a while. I think we have a limit of 5000 files, after which we give up, but maybe that is too high...

We could probably at some point move to using threads for the parsing, I guess...

I think the real question here is: is the parsing just taking up all that time, _or_ is there some further bug where the LS process ends up in an endless loop... @ZacLN any ideas how we might figure that out?

**Comment by davidanthoff** [comment]
@NHDaly we fixed an infinite loop in the LS on `master`, so that should ship soon.

Should we close this issue for now, and if this comes up again in newer versions we can reopen?

**Comment by NHDaly** [comment]
👍 i'm fine closing this for now. I fixed it locally just by moving the .julia out of the local directory. i can try putting it back in if you want me to to check if it's fixed? :) Lemme know!

**Comment by davidanthoff** [comment]
Yes, would be great if you could check this once the new release is out! Fingers crossed I'll release tomorrow.

**Comment by LeoGalPras** [comment]
Should this be corrected by now?? I am running version 0.15.29, julia v1.4, in Ubuntu 20.04, and I face a similar issue, not sure if it's the same though. After opening several julia files (not even tried running them) a process starts using 100% CPU for one core (if I open another folder and do the same, the same happens with another core), htop tells me its due to "/home/my_user/.vscode/extensions/julialng.language-julia.0,15.29/scripts/languageserver/packages/SymbolServer/src/server.jl" . This process remains open even after closing VSCode, I have to "killall julia" to get rid of it.

**Comment by davidanthoff** [comment]
@LeoGalPras What you are seeing is distinct. You are seeing the indexing server. That should be (hopefully) a one-time cost, once the indexing is finished for the bulk of your packages, the indexing service should only occasionally spin up again.

**Comment by LeoGalPras** [comment]
I something similar to what you say in another issue. But for me it never ends, there have been instances when I'm working for hours and this process never seems to end, it just keeps hogging CPU power, it's not only in the startup.

**Comment by lewisl** [comment]
Yet another reason to allow DISABLE LINTER.

PLEASE!!!!!

I don't have a julia linter in Sublime Text.  I can edit code.  Really we don't need this at all since it is WAY slow and generates hundreds of false positives.

Putting the Atom/Juno debugger in VS Code is a higher priority than any Linter, but even this is not so high priority.  Step debugger helps when <my> brain is broken and I don't understand my own <bad> code.  But, it is really a last resort.

**Comment by davidanthoff** [comment]
> Yet another reason to allow DISABLE LINTER.

There is an option for that.

> Putting the Atom/Juno debugger in VS Code is a higher priority than any Linter, but even this is not so high priority.

We've been shipping a debugger as part of the extension for a couple of months now.

**Comment by lewisl** [comment]
I’ll give it a “whirl”  (ha…)

**Comment by ptoche** [comment]
I'm seeing this too, why would indexing use 100% of CPU? On my 2018 MacBook Pro the fan is losing it.

<img width="839" alt="a" src="https://user-images.githubusercontent.com/6339274/117589067-4a410180-b0dc-11eb-9354-9374ef2b4be4.png">

**Comment by NHDaly** [comment]
Since this seems to bother and confuse many people, would it make sense to introduce regularly occurring artificial sleeps to reduce the CPU usage for this process? Since it's a background task anyway, there's no need for it to finish super quickly.

I know that it's actually more difficult to reduce a cpu-bound process's cpu usage in user-space than at the OS-level, but it might be worth the complexity to make this more user friendly?

**Comment by lewisl** [comment]
What does the process due?   Is it pre-compiling packages or indexing symbols in all of the packages? That seems a lengthy process indeed. Does it only index packages used by the current project or file?

- Lewis

On May 22, 2021, at 6:21 PM, Nathan Daly ***@***.***> wrote:

﻿

Since this seems to bother and confuse many people, would it make sense to introduce regularly occurring artificial sleeps to reduce the CPU usage for this process? Since it's a background task anyway, there's no need for it to finish super quickly.

I know that it's actually more difficult to reduce a cpu-bound process's cpu usage in user-space than at the OS-level, but it might be worth the complexity to make this more user friendly?

**Comment by pjgorski** [comment]
Such a process is created each time I start VSCode, even if I am not planning to use Julia but some other language. 
In my case, it doesn't take 100% CPU, but still, such a process should not be created without reason.

**Comment by pfitzseb** [comment]
It's created when the Julia extensions gets initialized, which only happens when you have Julia files in your workspace or use any command provided by the extension.

**Comment by pjgorski** [comment]
In my case, I open VSCode and the most recent project does not have a single .jl file (most likely no Julia notebooks as well), and all the opened files are Python notebooks or Python files.

**Comment by pfitzseb** [comment]
Yeah, I think the extension is also activated when you open jupyter notebooks to contribute our kernel. Maybe that's the case for you?
